Servers/UI/OJS.Servers.Ui/ClientApp/src/components/contests/contest-categories/ContestCategories.tsx

```diff
@@ -16,61 +18,198 @@ import styles from './ContestCategories.module.scss';
+
+            setOpenedStrategyFilters([
+                ...openedStrategyFilters,
+                { categoryId: id, strategies: strategyFiltersToAdd },
+            ]);
 
-        if (isNil(filter)) {
-            return;
-        }
+            setOpenedStrategyFilter({ categoryId: id, strategies: strategyFiltersToAdd });
+        },
+        [ getStrategyFiltersToAdd, openedStrategyFilters ],
+    );
 
-        onCategoryClick(filter);
-        updateBreadcrumb(category, categoriesFlat);
-    }, [ possibleFilters, categoriesFlat, onCategoryClick, updateBreadcrumb ]);
+    const removeOldStrategyFilters = useCallback(
+        (id: string) => {
+            const newFilterProps = openedStrategyFilters.filter((osf) => osf.categoryId !== id);
+
+            setOpenedStrategyFilters(newFilterProps);
+        },
+        [ openedStrategyFilters ],
+    );
+
+    useEffect(
+        () => {
+            if (isEmpty(openedStrategyFilters)) {
+                clearBreadcrumb();
+                setStrategyFilters([]);
+            } else {
+                setStrategyFilters(openedStrategyFilter.strategies);
+            }
+        },
+        [ clearBreadcrumb, openedStrategyFilter.strategies, openedStrategyFilters, setStrategyFilters ],
+    );
+
+    const updateStrategyFilters = useCallback(
+        (id: string, node: ICategoryStrategiesTypes) => {
+            const strategyFilterToUpdate = openedStrategyFilters.find(({ categoryId }) => categoryId === id);
+
+            (isNil(strategyFilterToUpdate)
+                ? addNewStrategyFilters
+                : removeOldStrategyFilters)(id, node);
+
+            setPrevCategoryId(id);
+        },
+        [ addNewStrategyFilters, openedStrategyFilters, removeOldStrategyFilters ],
+    );
+
+    const handleTreeLabelClick = useCallback(
+        (node: ICategoryStrategiesTypes) => {
+            const filter = possibleFilters.find(({ value }) => value.toString() === node.id.toString());
+            const category = categoriesFlat.find(({ id }) => id.toString() === node.id.toString());
+
+            if (isNil(filter)) {
+                return;
+            }
+
+            updateBreadcrumb(category, categoriesFlat);
+            updateStrategyFilters(node.id, node);
+            selectCurrentCategoryId(node.id);
+
+            onCategoryClick(filter);
+        },
+        [ categoriesFlat, onCategoryClick, possibleFilters, updateBreadcrumb, updateStrategyFilters ],
+    );
+
+    useEffect(
+        () => {
+            const currentNode = getCurrentNode(defaultSelected);
+
+            if (isNil(currentNode)) {
+                return;
+            }
+
+            selectCurrentCategoryId(currentNode?.id);
+        },
+        [ defaultSelected, getCurrentNode ],
+    );
+
+    useEffect(
+        () => {
+            if (currentCategoryId && currentCategoryId !== prevCategoryId) {
+                const categoryId = getCategoryId('', currentCategoryId);
```

Comment:
This use effect is too heavy. Please move the logic into a function. Also its a bit confusing, we get the category id but we already have a category id and we pass an empty value?